gapic/src/main/com/google/gapid/util/Logging.java

```diff
@@ -41,40 +42,80 @@
   /**
    * Possible log level flag values.
    */
-  public static enum LogLevel {
-    OFF(Level.OFF, "Emergency", "F"), ERROR(Level.SEVERE, "Error", "E"),
-    WARNING(Level.WARNING, "Warning", "W"), INFO(Level.INFO, "Info", "I"),
-    DEBUG(Level.FINE, "Debug", "D"), ALL(Level.ALL, "Debug", "V");
+  public enum LogLevel {
+    OFF(Level.OFF, Log.Severity.Fatal, "Fatal", "F"),
+    ERROR(Level.SEVERE, Log.Severity.Error, "Error", "E"),
+    WARNING(Level.WARNING, Log.Severity.Warning, "Warning", "W"),
+    INFO(Level.INFO, Log.Severity.Info, "Info", "I"),
+    DEBUG(Level.FINE, Log.Severity.Debug, "Debug", "D"),
+    ALL(Level.ALL, Log.Severity.Debug, "Debug", "V");
 
     public final Level level;
+    public final Log.Severity severity;
     public final String gapisLevel;
     public final String gapirLevel;
 
-    private LogLevel(Level level, String gapisLevel, String gapirLevel) {
+    LogLevel(Level level, Log.Severity severity, String gapisLevel, String gapirLevel) {
       this.level = level;
+      this.severity = severity;
       this.gapisLevel = gapisLevel;
       this.gapirLevel = gapirLevel;
     }
+
+    static LogLevel fromLevel(Level level) {
+      int intLevel = level.intValue();
+      switch (intLevel) {
+        case 1000: return ERROR;
+        case  900: return WARNING;
+        case  800: case  700: return INFO;
+        case  500: case  400: case  300: return DEBUG;
+        default:
+          if (intLevel > 1000) {
+            return ERROR;
+          } else if (intLevel > 900) {
+            return WARNING;
+          } else if (intLevel > 700) {
+            return INFO;
+          } else {
+            return DEBUG;
+          }
+      }
+    }
+  }
+
+  /**
+   * Listener is the interface implemented by types that want to listen to log messages.
+   */
+  public interface Listener {
+    void onNewMessage(Log.Message message);
   }
 
   public static final Flag<LogLevel> logLevel = Flags.value(
       "logLevel", LogLevel.INFO, "Logging level [OFF, ERROR, WARNING, INFO, DEBUG, ALL].");
   public static final Flag<String> logDir = Flags.value(
       "logDir", System.getProperty("java.io.tmpdir"), "Directory for log files.");
 
-  private static final MemoryHandler memoryHandler = new MemoryHandler() {{
+  private static final int BUFFER_SIZE = 1000;
+
+  private static final RingBuffer buffer = new RingBuffer(BUFFER_SIZE);
+
+  private static final Handler handler = new Handler() {{
```

Comment:
You no longer need to hang on to this reference, so just create the handler in the init() method and add it to the root logger.